src/main/java/com/google/finapp/SpannerDao.java

```diff
@@ -16,133 +16,183 @@
 
 import com.google.cloud.ByteArray;
 import com.google.cloud.spanner.*;
-import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.inject.Inject;
 
 import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Arrays;
 
 final class SpannerDao {
 
-  private final DatabaseClient databaseClient;
+  private final String projectId;
+  private final String databaseId;
+  private final String instanceId;
+  private final String connectionUrl;
 
   @Inject
-  SpannerDao(DatabaseClient databaseClient) {
-    this.databaseClient = databaseClient;
+  SpannerDao(@ArgsModule.SpannerProjectId String spannerProjectId,
+      @ArgsModule.SpannerInstanceId String spannerInstanceId,
+      @ArgsModule.SpannerDatabaseId String spannerDatabaseId) {
+    this.projectId = spannerProjectId;
+    this.databaseId = spannerDatabaseId;
+    this.instanceId = spannerInstanceId;
+    // use this URL to connect to the emulator
+    this.connectionUrl = String.format(
+        "jdbc:cloudspanner://localhost:9010/projects/%s/instances/%s/databases/%s;usePlainText=true",
+        projectId, instanceId, databaseId);
+    // use this URL to connect to Cloud Spanner
+    // this.connectionUrl = String.format(
+    //     "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
+    //     projectId, instanceId, databaseId);
   }
 
-  void createCustomer(ByteArray customerId, String name, String address) throws SpannerException {
-    databaseClient.write(
-        ImmutableList.of(
-            Mutation.newInsertBuilder("Customer")
-                .set("CustomerId")
-                .to(customerId)
-                .set("Name")
-                .to(name)
-                .set("Address")
-                .to(address)
-                .build()));
+  void createCustomer(ByteArray customerId, String name, String address) throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
```

Comment:
I'd expect the connection to be initialization once in the constructor and re-used for all classes

Reply:
Looking into this now. All the JDBC examples I saw followed a very similar format of establishing connections and statements at the start of functions similar to this, but this could just be because they were small sample snippets. 

Establishing the connection like this is also used in the [Cloud Spanner JDBC examples](https://github.com/GoogleCloudPlatform/java-docs-samples/tree/9defa78e095878d74b727f1e9fb6a74f06f0f12f/spanner/jdbc), which again can be just a result of showing simple individual function examples. 

But, yes, this makes sense. I'll reuse the connection.

Reply:
Actually, is it best practice to hold the same connection? This would mean that the connection would be used for the entire life of the server, since a singular SpannerDao instance is used. If this is the preferred practice, I'll implement it @voulg

Reply:
"The open-source JDBC driver handles pooling internally..." "These pool connections are re-used..." via https://cloud.google.com/spanner/docs/open-source-jdbc#do_i_need_to_provide_my_own_connection_pool (thanks @haseebajmal)